test/metabase/lib/join_test.cljc

```diff
@@ -1569,3 +1568,176 @@
+      (binding [lib.metadata.calculation/*display-name-style* :long]
+        (is (= [["TITLE" "Orders__TITLE" "TITLE" "Orders → Title"]
+                ["sum"   "Orders__sum"   "sum"   "Orders → Sum of Quantity"]]
+               (map (juxt :name :lib/desired-column-alias :lib/source-column-alias :display-name)
+                    (lib.join/join-fields-to-add-to-parent-stage
+                     query -1 join {:unique-name-fn (lib.util/unique-name-generator), :include-remaps? true}))))))))
+
+(deftest ^:parallel remapping-in-joins-test-3
+  (testing "join-fields-to-add-to-parent-stage should include remapped columns"
+    (let [mp    (lib.tu/remap-metadata-provider
+                 meta/metadata-provider
+                 (meta/id :orders :product-id) (meta/id :products :title))
+          query (lib/query
+                 mp
+                 (lib.tu.macros/mbql-query products
+                   {:joins    [{:source-query {:source-table $$orders}
+                                :alias        "Orders"
+                                :condition    [:= $id &Orders.orders.product-id]
+                                :fields       [&Orders.orders.product-id]}]
+                    :fields   [$title $category]}))
+          join (first (lib/joins query -1))]
+      (binding [lib.metadata.calculation/*display-name-style* :long]
+        (is (= [["PRODUCT_ID" "Orders__PRODUCT_ID" "PRODUCT_ID" "Orders → Product ID"]
+                ;; should get added because it is a remap
+                ["TITLE"      "Orders__TITLE"      "TITLE"      "Orders → Title"]]
+               (map (juxt :name :lib/desired-column-alias :lib/source-column-alias :display-name)
+                    (lib.join/join-fields-to-add-to-parent-stage
+                     query -1 join {:unique-name-fn (lib.util/unique-name-generator), :include-remaps? true}))))))))
+
+(deftest ^:parallel remapping-in-joins-duplicates-test
+  (testing "Remapped columns in joined source queries should not append duplicates (QUE-1410)"
+    (let [mp    (lib.tu/remap-metadata-provider
+                 meta/metadata-provider
+                 (meta/id :orders :product-id) (meta/id :products :title))
+          query (lib/query
+                 mp
+                 (lib.tu.macros/mbql-query products
+                   {:joins    [{:source-query {:source-table $$orders
+                                               :breakout     [$orders.product-id]
+                                               :aggregation  [[:sum $orders.quantity]]}
+                                :alias        "Orders"
+                                :condition    [:= $id &Orders.orders.product-id]
+                                ;; we can get title since product-id is remapped to title
+                                :fields       [&Orders.orders.product-id
+                                               &Orders.products.title]}]
+                    :fields   [$title $category]}))
```

Comment:
I'm scratching my head a bit at the density of `lib.tu.macros/mbql-query` in the net-new `lib` tests in this PR. They stand out because there's very little of that otherwise in the lib tests.

Some of them are testing specific or broken things that are hard to create with the lib API, so fair enough. But plenty, including this one, seem to be perfectly normal except that it's a PITA to populate the `:fields` lists, or more broadly to pick out columns from the query.

I don't think there's anything to change for this PR, but it has me thinking that we're missing a test helper. If we can have a usable shorthand for MBQL 4 we should be able to get the same for MBQL 5, without sacrificing the high-level nature of the threadable lib API.

Reply:
Most of these are copied and slightly modified from failing QP tests so we could have a version in Lib that makes sure it does what it's supposed to without involving the QP. So I didn't write most of these from scratch